include/continuable/detail/support/asio.hpp

```diff
@@ -0,0 +1,108 @@
+                                   v4.0.0
+
+  Copyright(c) 2015 - 2019 Denis Blank <denis.blank at outlook dot com>
+
+  Permission is hereby granted, free of charge, to any person obtaining a copy
+  of this software and associated documentation files(the "Software"), to deal
+  in the Software without restriction, including without limitation the rights
+  to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+  copies of the Software, and to permit persons to whom the Software is
+  furnished to do so, subject to the following conditions :
+
+  The above copyright notice and this permission notice shall be included in
+  all copies or substantial portions of the Software.
+
+  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  SOFTWARE.
+**/
+
+#ifndef CONTINUABLE_DETAIL_ASIO_HPP_INCLUDED
+#define CONTINUABLE_DETAIL_ASIO_HPP_INCLUDED
+
+#if defined(ASIO_STANDALONE)
+#include <asio/async_result.hpp>
+#include <asio/error_code.hpp>
+#include <asio/system_error.hpp>
+
+#define CTI_ASIO_NAMESPACE_BEGIN namespace asio {
+#define CTI_ASIO_NAMESPACE_END }
+#else
+#include <boost/asio/async_result.hpp>
+#include <boost/system/error_code.hpp>
+#include <boost/system/system_error.hpp>
+
+#define CTI_ASIO_NAMESPACE_BEGIN namespace boost { namespace asio {
+#define CTI_ASIO_NAMESPACE_END }}
+#endif
+
+#include <utility>
+
+namespace cti {
+namespace detail {
+namespace asio {
+
+#if defined(ASIO_STANDALONE)
+using error_code_t = ::asio::error_code;
+using exception_t = ::asio::system_error;
+#else
+using error_code_t = ::boost::system::error_code;
+using exception_t = ::boost::system::system_error;
+#endif
+
+// Binds `promise` to the first argument of a continuable resolver, giving it
+// the signature of an ASIO handler.
+template <typename Promise>
+auto promise_resolver_handler(Promise&& promise) noexcept {
+  return [promise = std::forward<Promise>(promise)](
+      error_code_t e, auto&&... args) mutable noexcept {
+    if (e) {
+      promise.set_exception(std::make_exception_ptr(exception_t(std::move(e))));
+    } else {
+      promise.set_value(std::forward<decltype(args)>(args)...);
+    }
+  };
+}
+
+// Type deduction helper for `Result` in `cti::make_continuable<Result>`
+template <typename Signature>
+struct continuable_result;
+
+template <>
+struct continuable_result<void(error_code_t)> {
+  using type = void;
+};
+
+template <>
+struct continuable_result<void(error_code_t const&)>
+    : continuable_result<void(error_code_t)> {};
+
+template <typename T>
+struct continuable_result<void(error_code_t, T)> {
```

Comment:
Is asio asynchronous result count capped to one parameter only?
Or do we need to provide additional specializations for multiple asynchronous results,
such as `void(error_code_t, int, float, char)`?

Reply:
Good question! In theory there is no cap on parameter count. In practice, you can look at ASIO type requirements (Ctrl + F "Handler" [here](https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference.html)), and all of them look like either `void(error_code_t)` or `void(error_code_t, T)`. 

In theory (again) a user could define a novel class of async operation whose completion handler requires more than one non-error parameter. Adding support for this is not hard and probably harmless. Keeping the `void(ec)` and `void(ec, T)` specializations as is, should I make it so that `void(ec, T, Ts...)` will deduce to `std::tuple<T, Ts...>` if `sizeof...(Ts) > 1` ?

Reply:
Oh wait a minute, I was reviewing the semantics of `continuable` with multiple result types--it seems it is just `continuable<Ts...>`, no tuple. I think this is still doable, it may just take some massaging. Up for debate how useful this will be to most people, but I'm going to ask around about if more parameters is indeed valid

Reply:
I think multiple results should probably be covered since in theory you are allowed to implement your own `io_object` like `deadline_timer` which could possibly provide async methods that return more than one asynchronous result and the fact that asio doesn't require it by now doesn't mean they could do that in the future.